### ISSUE TITLE
Invalid codegen around Proc accepting recursive alias

### ISSUE THREAD

**Issue by jhass** [crystal-lang/crystal#1485]
```cr
alias Alias = String|Int32|Array(Alias)

class Definition
  def initialize(&block : Proc(Alias, Bool | String))
    @validator = block
  end

  def validate
    @validator.call("foo" as Alias)
  end
end

Definition.new { "foo" }
Definition.new { true }.validate
```
```
==5554== Process terminating with default action of signal 11 (SIGSEGV)
==5554==  General Protection Fault
==5554==    at 0x570D238: GC_malloc (in /usr/lib/libgc.so.1.0.3)
==5554==    by 0x4256DDF: ???
==5554==    by 0x570D20C: GC_malloc (in /usr/lib/libgc.so.1.0.3)
==5554==    by 0x100000152: ???
==5554==    by 0x4259EFF: ???
==5554==    by 0x11FA08: *Definition::new<&(Alias -> Bool)>:Definition (bug.cr:6)
==5554==    by 0x10F6D7: __crystal_main (bug.cr:14)
==5554==    by 0x11147F: main (macro94846387057696.cr:9)
```

**Comment by asterite** [comment]
@jhass Seems to be fixed at 0.17.3, right?

**Comment by jhass** [comment]
I guess so, doesn't reproduce for me either. I don't remember where that's from, probably a reduction of somebody else's code.
